server/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java

```diff
@@ -381,6 +385,90 @@ public void onFailure(Exception e) {
         assertThat(failure.get().getMessage(), equalTo(catastrophicError.getMessage()));
     }
 
+    public void testCacheRemoteShardFailed() throws Exception {
+        final String index = "test";
+        setState(clusterService, ClusterStateCreationUtils.stateWithActivePrimary(index, true, randomInt(5)));
+        ShardRouting failedShard = getRandomShardRouting(index);
+        long primaryTerm = clusterService.state().metaData().index(index).primaryTerm(failedShard.id());
+        boolean markAsStale = randomBoolean();
+        int numListeners = between(1, 100);
+        CountDownLatch latch = new CountDownLatch(numListeners);
+        for (int i = 0; i < numListeners; i++) {
+            shardStateAction.remoteShardFailed(failedShard.shardId(), failedShard.allocationId().getId(),
+                primaryTerm + 1, markAsStale, "test", getSimulatedFailure(), new ShardStateAction.Listener() {
```

Comment:
why primaryTerm + 1? just a random but fixed value would do? (we never run this task)

Reply:
I pushed https://github.com/elastic/elasticsearch/pull/31313/commits/496a6bdfcf011d35a03947984ba618855ee3eca7. I copied this from an existing test.